structuredheadings/plugin.js

```diff
@@ -1,67 +1,139 @@
+      var setCommandState = function (command, state) {
         switch (state) {
         case "on":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_ON);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_ON);
           break;
         case "off":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_OFF);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_OFF);
           break;
         case "disabled":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_DISABLED);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_DISABLED);
           break;
         }
 
       };
 
+      var isNumbered = function (element) {
+        if (element.hasClass("autonumber")) {
+          return true;
+        } else {
+          return false;
+        }
+      };
+
       // The command to add an autonumbering class to selection
       editor.addCommand("autoNumberHeading", {
         allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
-        requiredContent: allowedElements.join(";"),
         startDisabled: true,
         exec: function () {
           var element = editor.getSelection().getStartElement();
 
-          if (!element.hasClass("autonumber")) {
+          if (!isNumbered(element)) {
             element.addClass("autonumber");
-            setCommandState("on");
+            setCommandState("autoNumberHeading", "on");
           } else {
             element.removeClass("autonumber");
-            setCommandState("off");
+            setCommandState("autoNumberHeading", "off");
           }
         }
       });
 
+      editor.addCommand("matchHeading", {
+        startDisabled: true,
+        exec: function () {
+          var element = editor.getSelection().getStartElement();
+          var style;
+          //find previous element that matches allowedElements
+          var previousHeader = element.getPrevious(function (node) {
+            if (allowedElements.indexOf(node.getName()) >= 0) {
+              return true;
+            } else {
+              return false;
+            }
+          });
+
+          // if already in header, set back to p
+          if (allowedElements.indexOf(element.getName()) >= 0) {
+            //eslint-disable-next-line new-cap
+            style = new CKEDITOR.style({ element: "p" });
+            editor.applyStyle(style);
+
+          // else get previous element style (type) and apply to selection
+          } else if (previousHeader) {
+            //eslint-disable-next-line new-cap
+            style = new CKEDITOR.style({ element: previousHeader.getName()});
+            // if previous was numbered, set the new  one to numbered also
+            if (isNumbered(previousHeader)) {
+              //eslint-disable-next-line new-cap
+              style = new CKEDITOR.style({ element: previousHeader.getName(),
+                attributes: {"class": "autonumber"}});
```

Comment:
as we discussed - documenting the discussion - it was more of a pondering thing.  It doesn't really matter, there is a minor efficiency gain in that we can save constructor calls if people frequently use > 6 styles, but this is so minor it's not worth thinking about